client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -20,266 +20,1049 @@
-		let mut inner = self.inner.write();
+	/// If the global number of blocks has been reached this method
+	/// will remove all subscriptions that have blocks older than the
+	/// specified pin duration.
+	///
+	/// If after removing all subscriptions that exceed the pin duration
+	/// there is no space for pinning a new block, then all subscriptions
+	/// are terminated.
+	///
+	/// Returns true if the given subscription is also terminated.
+	fn ensure_block_space(&mut self, request_sub_id: &str) -> bool {
+		if self.global_blocks.len() < self.global_max_pinned_blocks {
+			return false
+		}
 
-		if inner.blocks.len() == inner.max_pinned_blocks {
-			// We have reached the limit. However, the block can be already inserted.
-			if inner.blocks.contains(&hash) {
-				return Ok(false)
-			} else {
+		// Terminate all subscriptions that have blocks older than
+		// the specified pin duration.
+		let now = Instant::now();
+
+		let to_remove: Vec<_> = self
+			.subs
+			.iter_mut()
+			.filter_map(|(sub_id, sub)| {
+				let sub_time = sub.find_oldest_block_timestamp();
+				// Subscriptions older than the specified pin duration should be removed.
+				let should_remove = match now.checked_duration_since(sub_time) {
+					Some(duration) => duration > self.local_max_pin_duration,
+					None => true,
+				};
+				should_remove.then(|| sub_id.clone())
+			})
+			.collect();
+
+		let mut is_terminated = false;
+		for sub_id in to_remove {
+			if sub_id == request_sub_id {
+				is_terminated = true;
+			}
+			self.remove_subscription(&sub_id);
+		}
+
+		// Make sure we have enough space after first pass of terminating subscriptions.
+		if self.global_blocks.len() < self.global_max_pinned_blocks {
+			return is_terminated
+		}
+
+		// Sanity check: cannot uphold `chainHead` guarantees anymore. We have not
+		// found any subscriptions that have older pinned blocks to terminate.
+		let to_remove: Vec<_> = self.subs.keys().map(|sub_id| sub_id.clone()).collect();
+		for sub_id in to_remove {
+			if sub_id == request_sub_id {
+				is_terminated = true;
+			}
+			self.remove_subscription(&sub_id);
+		}
+		return is_terminated
+	}
+
+	fn pin_block(
+		&mut self,
+		sub_id: &str,
+		hash: Block::Hash,
+	) -> Result<bool, SubscriptionManagementError> {
+		let Some(sub) = self.subs.get_mut(sub_id) else {
+			return Err(SubscriptionManagementError::SubscriptionAbsent)
+		};
+
+		// Block was already registered for this subscription and therefore
+		// globally tracked.
+		if !sub.register_block(hash) {
+			return Ok(false)
+		}
+
+		// Ensure we have enough space only if the hash is not globally registered.
+		if !self.global_blocks.contains_key(&hash) {
+			// Subscription ID was terminated while ensuring enough space.
+			if self.ensure_block_space(sub_id) {
 				return Err(SubscriptionManagementError::ExceededLimits)
 			}
 		}
 
-		Ok(inner.blocks.insert(hash))
+		self.global_pin_block(hash)?;
```

Comment:
maybe this `global_pin_block` action can be moved within the `!self.global_blocks.contains_key(&hash)` scope

Reply:
No I  was thinking about :
```
if !self.global_blocks.contains_key(&hash) {
     if self.ensure_block_space(sub_id) {
         return Err...
     }
     self.global_pin_block(hash)?;
}
```

Reply:
Ahh I see, the `global_pin_block` function name is a bit misleading, under the hood it does backend pinning if the block does not exist or increments the internal counter if the block exists. 
This is similar to the `global_unpin_block`.

Will renames those to `global_register_block` and `global_unregister_block` similar to `SubscriptionState` naming, and add a couple of documentation :D